### DISCUSSION TITLE
Selecting missing residues

### DISCUSSION THREAD

**Discussion by Marwa-Da** [molstar/molstar#1009] [Q&A]
Hi,

Is it possible to change the code implementation so that the missing residues could be selected in the sequence panel?

**Comment by Marwa-Da** [comment]
I digged deeply into the code files until I reached [this line](https://github.com/molstar/molstar/blob/2f0230dc8449b8f91af8eff9cb7817dded71f50b/src/mol-plugin/behavior/dynamic/representation.ts#L169) but couldn't figure out yet where the bindings are assigned

**Comment by dsehnal** [comment]
This is currently not possible, as the `Loci` for the unobserved residues are not implemented at all.

I can see two options:

- Write your own version of the `Sequence` component that raises a custom event (can just create a new `BehaviorSubject` for this say in `plugin.customState.mySeqEvent = new BehaviorSubject(...)` for the unobserved residues
- Define a new kind of [`Loci`](https://github.com/molstar/molstar/blob/master/src/mol-model/loci.ts#L67) that is able to represent a pointer to the unobserved residue. This would likely require more planning and making sure this doesn't break anything downstream (generally, it shouldn't)

**Reply by arose** [reply]
For option2 I would prefer having a general sequence loci (#58).

**Reply by Marwa-Da** [reply]
@arose Would you please elaborate a bit ?

**Reply by Marwa-Da** [reply]
@dsehnal Really appreciate your quick answers!

**Reply by dsehnal** [reply]
> For option2 I would prefer having a general sequence loci (#58).

Yea, that's what it would boil down to, should have been more explicit about it.

**Reply by arose** [reply]
> @arose Would you please elaborate a bit ?

rough sketch, @Marwa-Da feel free to make a PR following it @dsehnal what do you think?

- Define a new Sequence Loci in `src\mol-model\sequence\sequence.ts` similar to the Structure Loci in `src\mol-model\structure\structure\structure.ts`
- Interface of the Sequence Loci: `{ kind: 'sequence-loci', sequence: Sequence, indices: OrderedSet<number> }`
- Where indices are what is returned by Sequence.Base.index()
- Add the new Loci to https://github.com/molstar/molstar/blob/master/src/mol-model/loci.ts#L67 and helpers in that file if it makes sense
- In src\mol-plugin-ui\sequence\sequence.tsx we emit Sequence Loci instead o StructureElement Loci
- We then need a way to convert a Sequence Loci to Structure Element Loci because the highlight/select manager only works for Structure Element Loci. This needs to be done at the application level because there is are no pointers from a Sequence to a Structure object. What we want is to convert a Sequence loci to a Structure Element Loci for all root structures (`plugin.state.data.select(StateSelection.Generators.rootsOfType(PluginStateObject.Molecule.Structure))`) that contain the Sequence, specifically the units that have the matching sequence as part of their Model (`unit.model.sequence`). Maybe `src\mol-plugin-state\manager\interactivity.ts#InteractivityManager.LociMarkManager.normalizedLoci` is a good place to do it. We may want to cache the mapping between sequences and structures; maybe similar to `src\mol-plugin\util\substructure-parent-helper.ts`.

**Reply by Marwa-Da** [reply]
Perfect! I'll start with it within two days. I'm eager to implement this, motivated by your collaborative approach in discussions and pull requests ^_^ .
